tensorflow/stream_executor/BUILD

```diff
@@ -51,6 +52,14 @@ cc_library(
     ] + if_static([":stream_executor_impl"]),
 )
 
+cc_header_only_library(
+    name = "stream_executor_headers_lib",
+    visibility = ["//visibility:public"],
+    deps = [
+        ":stream_executor",
+    ],
+)
```

Comment:
What problem is this new build rule solving?  I thought this was the purpose of //tensorflow/core:stream_executor_headers_lib.

Reply:
@jlebar Without the headers_lib, the following error will show up:
```
ERROR: /home/ubuntu/tensorflow/tensorflow/contrib/mpi_collectives/BUILD:40:1: undeclared inclusion(s) in rule '//tensorflow/contrib/mpi_collectives:python/ops/_mpi_ops.so':
this rule is missing dependency declarations for the following files included by 'tensorflow/contrib/mpi_collectives/kernels/mpi_ops.cc':
  'tensorflow/stream_executor/lib/statusor.h'
  'tensorflow/stream_executor/lib/status.h'
  'tensorflow/stream_executor/lib/error.h'
  'tensorflow/stream_executor/lib/stringpiece.h'
  'tensorflow/stream_executor/platform/port.h'
  'tensorflow/stream_executor/platform/logging.h'
  'tensorflow/stream_executor/lib/statusor_internals.h'
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

So I added the `stream_executor_headers_lib` to address the above `undeclared inclusion(s)` issue.

Reply:
> Without the headers_lib, the following error will show up:

That explains why you need to add a new dependency to `tensorflow/contrib/mpi_collectives/BUILD`.

My question is, can that dependency be `//tensorflow/core:stream_executor_headers_lib` instead of this new rule?  If not, why not?

Reply:
@jlebar With `//tensorflow/core:stream_executor_headers_lib` the following error will show up:
```
ERROR: /home/ubuntu/tensorflow/tensorflow/contrib/mpi_collectives/BUILD:40:1: Label '//tensorflow/core:stream_executor_headers_lib' is duplicated in the 'deps' attribute of rule 'python/ops/_mpi_ops_gpu'
ERROR: /home/ubuntu/tensorflow/tensorflow/contrib/mpi_collectives/BUILD:40:1: Label '//tensorflow/core:stream_executor_headers_lib' is duplicated in the 'deps' attribute of rule 'python/ops/_mpi_ops.so_check_deps'
ERROR: /home/ubuntu/tensorflow/tensorflow/contrib/mpi_collectives/BUILD:40:1: Label '//tensorflow/core:stream_executor_headers_lib' is duplicated in the 'deps' attribute of rule 'python/ops/_mpi_ops.so'
```

I am not sure about the error, though from `_mpi_ops_gpu` I guess `stream_executor` may have gpu and no-gpu version and that might be the conflict?

Reply:
Ah, indeed, tensorflow/core:stream_executor_headers_lib depends on CUDA stuff.  So the new build rule you added is different in this important way.

These build rules are extremely messed up.

I'm happy to take this for now, with the caveat that if and when we clean up the SE build rules, we may break you again.  I don't like that, but apparently we have no continuous builds for MPI.  :-/